relayer/processor/path_processor_internal.go

```diff
@@ -18,47 +19,90 @@ import (
 // i.e. a MsgConnectionOpenInit or a MsgChannelOpenInit should be broadcasted to start
 // the handshake if this key exists in the relevant cache.
 const (
-	preInitKey  = "pre_init"
-	preCloseKey = "pre_close"
+	preInitKey         = "pre_init"
+	preCloseKey        = "pre_close"
+	maxPacketsPerFlush = 10
 )
 
 // getMessagesToSend returns only the lowest sequence message (if it should be sent) for ordered channels,
 // otherwise returns all which should be sent.
 func (pp *PathProcessor) getMessagesToSend(
+	ctx context.Context,
 	msgs []packetIBCMessage,
 	src, dst *pathEndRuntime,
 ) (srcMsgs []packetIBCMessage, dstMsgs []packetIBCMessage) {
 	if len(msgs) == 0 {
 		return
 	}
+
 	if msgs[0].info.ChannelOrder == chantypes.ORDERED.String() {
-		// for packet messages on ordered channels, only handle the lowest sequence number now.
-		sort.SliceStable(msgs, func(i, j int) bool {
-			return msgs[i].info.Sequence < msgs[j].info.Sequence
-		})
-		firstMsg := msgs[0]
-		switch firstMsg.eventType {
-		case chantypes.EventTypeRecvPacket:
-			if dst.shouldSendPacketMessage(firstMsg, src) {
-				dstMsgs = append(dstMsgs, firstMsg)
+		eventMessages := make(map[string][]packetIBCMessage)
+
+		for _, m := range msgs {
+			eventMessages[m.eventType] = append(eventMessages[m.eventType], m)
+		}
+
+		for e, m := range eventMessages {
+			m := m
+			sort.SliceStable(m, func(i, j int) bool {
+				return m[i].info.Sequence < m[j].info.Sequence
+			})
+
+			if e == chantypes.EventTypeRecvPacket {
+				res, err := dst.chainProvider.QueryNextSeqRecv(ctx, 0, m[0].info.DestChannel, m[0].info.DestPort)
+				if err != nil {
+					dst.log.Error("Failed to query next sequence recv", zap.Error(err))
+					return
+				}
+
+				if m[0].info.Sequence != res.NextSequenceReceive {
+					dst.log.Error("Unexpected next sequence recv",
+						zap.String("channel_id", m[0].info.DestChannel),
+						zap.String("port_id", m[0].info.DestChannel),
+						zap.Uint64("expected", res.NextSequenceReceive),
+						zap.Uint64("actual", m[0].info.Sequence),
+					)
+					return
+				}
 			}
-		default:
-			if src.shouldSendPacketMessage(firstMsg, dst) {
-				srcMsgs = append(srcMsgs, firstMsg)
+
+			for i, msg := range m {
+				// only handle consecutive sequences on ordered channels
+				if i > 0 && msg.info.Sequence-1 != m[i-1].info.Sequence {
+					dst.log.Error("Packets are not consecutive",
+						zap.String("channel_id", m[0].info.DestChannel),
+						zap.String("port_id", m[0].info.DestChannel),
+						zap.Uint64("seq", msg.info.Sequence),
+						zap.Uint64("prior_seq", m[i-1].info.Sequence),
+					)
+					break
+				}
```

Comment:
This happened frequently during the neutron initial VSC packet backlog flush. The reason is that it now it will only attempt to flush a few packets at a time. If newer packets are observed from the chain when there is a backlog of packets to flush, they may not be consecutive. This is especially true for larger backlogs of packets.

Reply:
Ahh okay, I was having some trouble conceptualizing a case where this would happen. Thanks for breaking that down!